### ISSUE TITLE
Lazy context

### ISSUE THREAD

**Issue by 9999years** [winnow-rs/winnow#605]
### Please complete the following tasks

- [X] I have searched the [discussions](https://github.com/winnow-rs/winnow/discussions)
- [X] I have searched the [open](https://github.com/winnow-rs/winnow/issues) and [rejected](https://github.com/winnow-rs/winnow/issues?q=is%3Aissue+label%3AS-wont-fix+is%3Aclosed) issues

### winnow version

`master`

### Describe your use case

I'd like to add context to parsers without immediately allocating a `Vec` for them

### Describe the solution you'd like

Would it be possible to have a `.context_from(|| vec!["context1", "context2"])` function which only computes the context on error branches? Ideally this would be a middle-ground for performance between allocating the contexts up-front and omitting them entirely.

### Alternatives, if applicable

Not sure! I'm interested to hear discussion.

### Additional Context

See #421.

**Comment by epage** [comment]
Could you share your error type and context type?

**Comment by 9999years** [comment]
I'm using `PResult<ContextError<StrContext>>`, but I'll probably create my own context type like this so I can dynamically construct informative parse context messages:

```rust
#[derive(Debug, Clone)]
pub enum ParseContext {
    Description(Cow<'static, str>),
    Expected(Cow<'static, str>),
}
```

**Comment by epage** [comment]
`ContextError` doesn't immediately allocate but  does only pushes on error.

I'm confused by the example because you don't pass a `Vec` to `context` but a single context.

**Comment by 9999years** [comment]
> I'm confused by the example because you don't pass a `Vec` to `context` but a single context.

Yeah, sorry, to be clear I mean a method which computes multiple contexts up-front.

> `ContextError` doesn't immediately allocate but does only pushes on error.

You're right, thank you. I've looked at the code a bit more. But the `Parser::context` method does take the context up-front, which means that the cost of computing the context itself is always paid. Not much of an issue for `&'static str` contexts, but for anything else it could be expensive:

https://github.com/winnow-rs/winnow/blob/8674ed2c3f57963a4e96c56616ee0a35f58cd258/src/parser.rs#L646-L660

Maybe I'm just holding it wrong? Not sure.

**Comment by epage** [comment]
It would help if you gave a more complete example of what you are trying to accomplish.

**Comment by epage** [comment]
Without further details, I'm going to go ahead and close this.  When details become available and we can see a concrete change to be made, we can re-open this!
